[Overlay] Scrolllock utility

KenAJoh opened the pull request:
### Description

In some components like Drawer, Modal etc we need to "lock" the scroll on screen. There are some npm-packages that handle this: react-remove-scroll, react-scrolllock. But to keep dependencies lower, this solution is from BaseUi. 
The benefit of the more complex solution here compared to `useBodyScrollLock` is that it respects scrollbare-gutter, and avoids layout-shifts when opening Modal since it inserts `scrollbar-gutter: stable` when enabled. 

### Component Checklist 📝

- [ ] JSDoc
- [ ] Examples
- [ ] Documentation / Decision Records
- [ ] Storybook
- [ ] Style mappings (`@navikt/core/css/config/_mappings.js`)
- [ ] Component tokens (`@navikt/core/css/tokens.json`)
- [ ] CSS class deprecations (`@navikt/aksel-stylelint/src/deprecations.ts`)
- [ ] Exports (`@navikt/core/react/src/index.ts` and `@navikt/core/react/package.json`)
- [ ] New component? CSS import (`@navikt/core/css/index.css`)
- [ ] Breaking change? Update migration guide. Consider codemod.
- [ ] Changeset (Format: `<Component>: <gitmoji?> <Text>.` E.g. "Button: :sparkles: Add feature xyz.")

KenAJoh committed:
:sparkles: Added scrolllock utility

KenAJoh committed:
:tada: UseScrollLock story

KenAJoh committed:
:recycle: Use scrolllock utility in Modal

KenAJoh committed:
:memo: Changeset

HalvorHaugan commented on .changeset/true-ants-draw.md:
```diff
@@ -0,0 +1,6 @@
+"@navikt/ds-css": patch
+---
+
+Modal: Replace className placed on Body when Modal is Open, repaced scroll-locking with new internal utility.
```
1. Maybe better to write that className was removed (alternatively what it was replaced with).
2. Typo: "repaced".

HalvorHaugan commented on @navikt/core/css/modal.css:
```diff
@@ -1,7 +1,3 @@
-.navds-modal__document-body {
-  overflow: hidden;
-}
```
Hmm... Looks like this will be breaking for some🤔 https://github.com/search?q=org%3Anavikt+navds-modal__document-body+NOT+repo%3Anavikt%2Faksel&type=code

Should we...
- add the class to the stylelint rule?
- mention the class in the changeset?
- alert affected teams?

KenAJoh replied:
So i ended up just re-adding the bodyScrollLock util hook to support backwards compat.
- Removed the className with "aksel"-prefix since its not used by anyone yet, so not needed anymore.
- Still kept deprecations in stylelint, but just tagged it as a future deprecation.

So no need to manage anything right now since it will still work as before for those few who have used it.

HalvorHaugan commented on @navikt/core/react/src/modal/Modal.test.tsx:
```diff
@@ -31,33 +30,23 @@ describe("Modal", () => {
 
   test("should toggle body class", async () => {
```
```suggestion
  test("should toggle scroll lock", async () => {
```

HalvorHaugan commented on @navikt/core/react/src/modal/ModalUtils.ts:
```diff
@@ -28,48 +27,31 @@ export function getCloseHandler(
-      } else {
-        ownerDoc.body.classList.remove(BODY_CLASS, BODY_CLASS_LEGACY);
-      }
+      setIsOpen(!!modalRef.open);
```
Do we need `!!` ?

KenAJoh replied:
Nope

HalvorHaugan commented on @navikt/core/react/src/util/hooks/Hooks.stories.tsx:
```diff
@@ -0,0 +1,104 @@
+      >
+        <h2>Resize to go from no scroll to scroll</h2>
+        <div>{story()}</div>
+        <div style={{ width: 400, resize: "horizontal", overflow: "auto" }}>
```
I get no scrollbars here. Did you forget to set a static height?

KenAJoh replied:
Its more of a manual test where one reizes the text to get scrollbars, so can test both with and without scrollbars manually

HalvorHaugan replied:
You can only resize horizontally, so it's impossible to get scrollbars...

KenAJoh replied:
Might be misunderstanding each other here, but scroll works here:
<img width="1512" height="867" alt="Screenshot 2025-10-23 at 13 29 44" src="https://github.com/user-attachments/assets/2369101b-5a9f-47ad-a2bf-1035cd7d23b5" />

HalvorHaugan replied:
Doesn't work for me in Chrome or FF. Maybe an OS thing? Seems like a bug that you get a vertical scrollbar here, since the container should just expand vertically to fit the content.

HalvorHaugan replied:
Ignore previous comment. I misunderstood the purpose of the resizable div 🙈

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+    };
+
+    /* Handle `scrollbar-gutter` in Chrome when there is no scrollable content. */
+    const supportsStableScrollbarGutter =
```
I find this comment confusing since the code following it doesn't seem to actually handle anything.

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+    body.scrollTop = scrollTop;
+    body.scrollLeft = scrollLeft;
+    html.setAttribute("data-aksel-scroll-locked", "");
+    html.style.scrollBehavior = "unset";
```
Why not set `scrollBehavior` inside the `Object.assign` above?

KenAJoh replied:
Just keeps things consistent since `originalHtmlScrollBehavior` is a separate const

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+      parseFloat(bodyStyles.marginLeft) + parseFloat(bodyStyles.marginRight);
+
+    /*
+     * DOM writes:
+     * Do not read the DOM past this point!
+     */
+
+    Object.assign(html.style, {
+      scrollbarGutter: "stable",
+      overflowY:
+        !supportsStableScrollbarGutter &&
+        (isScrollableY || hasConstantOverflowY)
+          ? "scroll"
+          : "hidden",
+      overflowX:
+        !supportsStableScrollbarGutter &&
+        (isScrollableX || hasConstantOverflowX)
+          ? "scroll"
+          : "hidden",
+    });
+
+    Object.assign(body.style, {
+      position: "relative",
+      height:
+        marginY || scrollbarHeight
+          ? `calc(100dvh - ${marginY + scrollbarHeight}px)`
+          : "100dvh",
+      width:
+        marginX || scrollbarWidth
+          ? `calc(100vw - ${marginX + scrollbarWidth}px)`
+          : "100vw",
+      boxSizing: "border-box",
+      overflow: "hidden",
+      scrollBehavior: "unset",
+    });
```
Would be nice with comments explaining why each of these CSS properties are needed (except `overflow`). (Can put it on top of file or something if it's too messy to have it here.)

HalvorHaugan replied:
Is it possible to create a story that has the stuff that position/height/width is supposed to account for?

KenAJoh replied:
A little hard to test since the changes need to be on `body`, but i think these screenshots shows it nicely

This happens if we don't account for margin on Body. The height is then 100dvh + margin, causing the content to clip at the bottom of the page
<img width="965" height="827" alt="Screenshot 2025-10-24 at 12 26 07" src="https://github.com/user-attachments/assets/88a133f6-46c5-4132-b55e-0bdb0c0b5654" />

If we account for the margin properly, setting in this example 4rem margin on body will properly show 4 rem margin around the body-wrapper.
<img width="969" height="828" alt="Screenshot 2025-10-24 at 12 26 13" src="https://github.com/user-attachments/assets/62d0439e-6cdb-47a9-b7f9-e9d09451aecd" />

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+
+    Object.assign(html.style, {
+      scrollbarGutter: "stable",
+      overflowY:
+        !supportsStableScrollbarGutter &&
+        (isScrollableY || hasConstantOverflowY)
+          ? "scroll"
+          : "hidden",
+      overflowX:
+        !supportsStableScrollbarGutter &&
+        (isScrollableX || hasConstantOverflowX)
+          ? "scroll"
+          : "hidden",
```
So if gutter is not supported, we don't lock the scroll?

KenAJoh replied:
Yes, since we then can still preserve the scrollbar, while `overflow: hidden` is set on Body in that case locking the vertical scroll anyways.

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+    if (resizeRaf) {
+      cancelAnimationFrame(resizeRaf);
+    }
+    resizeRaf = requestAnimationFrame(lockScroll);
```
Do we need to use `requestAnimationFrame` here?

KenAJoh replied:
Since resize can often ends up changing layout, we want to wait to the next frame so that we can ensure lockScroll gets the most up-to-date data. If we do it instantly, we could end up with some layout trashing.

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+  acquire(referenceElement: Element | null) {
+    this.lockCount += 1;
+    if (this.lockCount === 1 && this.restore === null) {
+      this.timeoutLock.start(0, () => this.lock(referenceElement));
```
Why is the timeout needed?

KenAJoh replied:
Delays lock-handling so that other elements locking/unlocking in quick succession won't lead to layout trashing.

HalvorHaugan replied:
How does delaying prevent layout trashing?

KenAJoh replied:
Calling `getComputedStyle` causes a layout-flush, if we do this synchronously we get a read->write->read->write, but by using `timeoutLock.start` we cancel the first run if two locks try to lock at the same time.

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+
+    /*
+     * Avoid shift due to the default <body> margin. This does cause elements to be clipped
+     * with whitespace. Warn if <body> has margins?
```
How to reproduce this clipping?

KenAJoh replied:
Add `white-space: nowrap;` to the text in story
<img width="1116" height="310" alt="Screenshot 2025-10-21 at 13 45 33" src="https://github.com/user-attachments/assets/f38548e2-93b5-400d-ad91-e426e53577e0" />

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+let originalBodyStyles: Partial<CSSStyleDeclaration> = {};
+let originalHtmlScrollBehavior = "";
+
+const NOOP = () => {};
```
`NOOP` is used only one place. Does it need to be a const?

KenAJoh replied:
Nope, removed it

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+      return;
+    }
+
+    const isOverflowHiddenLock = isIOS || !hasInsetScrollbars(referenceElement);
```
Both locking methods use overflow:hidden in some way. I suggest renaming  "isOverflowHiddenLock" to e.g. "useBasicLock".

KenAJoh replied:
Called it `shouldUseBasicLock`

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+   * https://github.com/mui/base-ui/issues/1135
+   * This only applies changes in the state where element is:
+   * - No longer open
+   * - Is mounted (and probably animating out)
```
The line "This only applies..." is a bit confusing to me. It's not obvious what "changes" and "state" refers to. Maybe what's most relevant to explain is that !open + mounted probably means we are animating out?

```suggestion
   * mounted + not open = probably animating out
```
Might be better to put that right after the `if` in question though.

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+  const { enabled = true, mounted, open, referenceElement = null } = params;
+
+  /**
+   * https://github.com/mui/base-ui/issues/1135
```
Maybe add a short summary of the issue so that we don't have to open the link to understand/remember what it's about.

HalvorHaugan commented on @navikt/core/react/src/util/detectBrowser.ts:
```diff
@@ -0,0 +1,46 @@
+const isSafari = hasNavigator && /apple/i.test(navigator.vendor);
+
+const isWebKit =
+  typeof CSS === "undefined" || !CSS.supports
+    ? false
+    : CSS.supports("-webkit-backdrop-filter:none");
+
+const isIOS =
+  /*  */
+  nav.platform === "MacIntel" && nav.maxTouchPoints > 1
+    ? true
+    : /iP(hone|ad|od)|iOS/.test(nav.platform);
```
This could in theory stop working without warning. Can we live with that?

KenAJoh replied:
Yes i would say so. Since these are used for progressive enhancement, them breaking for some won't render the component unusable.

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,291 @@
+    /**
+     * On iOS, scroll locking does not work if the navbar is collapsed.
+     * Due to side-effects and bugs that arise on iOS, it must be researched extensively before
+     * being enabled to ensure it doesn't cause the following issues:
```
The first two sentences of this comment is a bit unclear. Unfortunately it's hard to explain why 😅🙈
- I think it's unclear whether the two sentences have something to do with each other, if that makes sense. I.e. does scroll lock not work b.c. it causes side-effects and bugs?
- Is the "side-effects and bugs that arise on iOS" related to `preventScrollStandard`, and that's why we only use `preventScrollBasic` in this case?

KenAJoh replied:
Change comment to 
```
    /**
     * On iOS, scroll locking does not work if the navbar is collapsed.
     * Due to the the following cases not working well with the standard scroll lock:
     * - Textboxes must scroll into view when focused, nor cause a glitchy scroll animation.
     * - The navbar must not force itself into view and cause layout shift.
     * - Scroll containers must not flicker upon closing a popup when it has an exit animation.
     *
     * We use a simpler scroll lock strategy to avoid these issues.
     */
```

HalvorHaugan commented on @navikt/core/react/src/util/hooks/Hooks.stories.tsx:
```diff
@@ -0,0 +1,104 @@
+        </h3>
+        <button onClick={() => setEmulateIosState((x) => !x)}>
+          Emulate WebKit-state: {`${emulateIosState}`}
+        </button>
```
Is it possible to reproduce the webkit issue in any of these stories?

KenAJoh replied:
No, since its only testable used together with an actual animation/transition. I removed the code for since its not testable.

HalvorHaugan commented on @navikt/core/react/src/util/hooks/Hooks.stories.tsx:
```diff
@@ -0,0 +1,104 @@
+      <div>
+        <h3>
+          When closing (!open && mounted), lazy clicks can end up selecting
+          whole page when transitioning to removing scroll-lock
```
Should mention that this only applies to WebKit, and explain what lazy clicks are (maybe link to the issue here as well).

HalvorHaugan commented on @navikt/core/react/src/util/hooks/Hooks.stories.tsx:
```diff
@@ -0,0 +1,104 @@
+          whole page when transitioning to removing scroll-lock
+        </h3>
+        <button onClick={() => setEmulateIosState((x) => !x)}>
+          Emulate WebKit-state: {`${emulateIosState}`}
```
I suggest renaming button to "Emulate mounted && !open" and `setEmulateIosState` accordingly.

KenAJoh committed:
:memo: Update changeset

KenAJoh committed:
:memo: Added className to list of styellint className deprecations

KenAJoh committed:
Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
:fire: removed double negative negation

KenAJoh committed:
:memo: Story-docs

KenAJoh committed:
:memo: Document body style overrides

KenAJoh committed:
:memo: Docs

KenAJoh committed:
:recycle: Cleanup redundant fn

KenAJoh committed:
:recycle: better naming

KenAJoh committed:
Update @navikt/core/react/src/util/hooks/useScrollLock.ts

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
:memo: Docs

KenAJoh committed:
:fire: removed test-case in story that cant work

KenAJoh committed:
Merge branch 'main' into overlay-scrolllock

KenAJoh committed:
:memo: Update docs

KenAJoh committed:
:memo: Update docs

KenAJoh committed:
:fire: removed comment

KenAJoh committed:
:tada: Re-added bodylockscroll support for backwards compat

KenAJoh committed:
:tada: Re-added class to css, removed updated aksel prefix css

KenAJoh committed:
Merge branch 'main' into overlay-scrolllock

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,318 @@
+    const scrollbarHeight = Math.max(0, win.innerHeight - html.clientHeight);
+
+    /*
+     * Avoid shift due to the default <body> margin. This does cause elements to be clipped
```
```suggestion
     * Avoid shift due to <body> margin. NB: This does cause elements to be clipped
```

HalvorHaugan commented on @navikt/core/react/src/util/hooks/useScrollLock.ts:
```diff
@@ -0,0 +1,318 @@
+     * Check support for stable scrollbar gutter to avoid layout shift when scrollbars appear/disappear.
+     */
+    const supportsStableScrollbarGutter =
+      typeof CSS !== "undefined" &&
+      CSS.supports?.("scrollbar-gutter", "stable");
+    /*
```
```suggestion
      CSS.supports?.("scrollbar-gutter", "stable");

    /*
```

KenAJoh committed:
Update @navikt/core/react/src/util/hooks/useScrollLock.ts

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
Update @navikt/core/react/src/util/hooks/useScrollLock.ts

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
:tada: Update templates for form navigation (#4194)

* :tada: Update templates for form navigation

* :bug: Revert to correct icon

* Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

* Update aksel.nav.no/website/components/website-modules/examples/__parts-inline/FormNavigationSimple.tsx

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

---------

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
:bug: Add templates to chromatic build

KenAJoh committed:
:tada: Update Accordion-header size (#4209)

* :tada: Update Accordion-header size

* Apply suggestion from @KenAJoh

KenAJoh committed:
:test_tube: Demo story

KenAJoh committed:
:test_tube: Demo story

HalvorHaugan committed:
iOS comment adjustments

HalvorHaugan commented on @navikt/core/react/src/modal/Modal.tsx:
```diff
@@ -144,6 +149,17 @@ export const Modal = forwardRef<HTMLDialogElement, ModalProps>(
       }
     }, [portalNode, open]);
 
+    useScrollLock({
+      enabled: isModalOpen,
+      mounted: isModalOpen,
+      open: isModalOpen,
+      referenceElement: modalRef.current,
+    });
```
I think we should have a changeset for this, e.g. "Modal: Improved scroll locking"

HalvorHaugan commented on @navikt/core/react/CHANGELOG.md:
Looks like you made an accidental change to this changelog, as well as the one on root 😅

HalvorHaugan committed:
Fix Chromatic bug in FormNavigationSimple.tsx

HalvorHaugan committed:
Tests: Fix warnings about act (#4218)

dependabot[bot] committed:
:arrow_up: Bump actions/download-artifact from 5 to 6 (#4225)

Bumps [actions/download-artifact](https://github.com/actions/download-artifact) from 5 to 6.
- [Release notes](https://github.com/actions/download-artifact/releases)
- [Commits](https://github.com/actions/download-artifact/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/download-artifact
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
:arrow_up: Bump actions/upload-artifact from 4 to 5 (#4224)

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 4 to 5.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
:arrow_up: Bump esbuild from 0.25.10 to 0.25.11 (#4222)

Bumps [esbuild](https://github.com/evanw/esbuild) from 0.25.10 to 0.25.11.
- [Release notes](https://github.com/evanw/esbuild/releases)
- [Changelog](https://github.com/evanw/esbuild/blob/main/CHANGELOG.md)
- [Commits](https://github.com/evanw/esbuild/compare/v0.25.10...v0.25.11)

---
updated-dependencies:
- dependency-name: esbuild
  dependency-version: 0.25.11
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

HalvorHaugan committed:
Adjust Slack feedback text (#4217)

HalvorHaugan committed:
Storybook: Fix story naming in Chromatic for templates (#4226)

HalvorHaugan committed:
Combobox: Fix issue with toggle list button (#4228)

github-actions[bot] committed:
Version Packages (#4214)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KenAJoh committed:
:memo: Changeset

KenAJoh commented on @navikt/core/react/CHANGELOG.md:
```diff
@@ -1846,6 +1846,7 @@
 ### Minor Changes
 
 - :recycle: Refaktorering og småfikser ([#2265](https://github.com/navikt/aksel/pull/2265))
+
```
```suggestion
```

KenAJoh committed:
Apply suggestion from @KenAJoh

KenAJoh commented on CHANGELOG.md:
```diff
@@ -1845,6 +1845,7 @@
 ### @navikt/ds-react
 
 - :recycle: Refaktorering og småfikser ([#2265](https://github.com/navikt/aksel/pull/2265))
+
```
```suggestion
```

KenAJoh committed:
Apply suggestion from @KenAJoh
